### PR TITLE
Throw an exception if a module requiring synchronization on lumi boundaries is used when concurrent lumis are enabled

### LINKED ISSUE #121: Change LogSystem to exception when trying to run a job with >= 2 concurrent lumis with an incompatible module
Exception message should tell clearly that the exception can be avoided by setting `process.options.numberOfConcurrentLuminosityBlocks = 1`.

Done in https://github.com/cms-sw/cmssw/issues/35326

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

Following the completion of #25090. This PR simply turns an existing warning to an exception. It also addresses framework unit tests that would fail because of the exception. It is possible that some unit tests in other packages would fail too (in which case I'll address those separately).

Resolves https://github.com/makortel/framework/issues/121

#### PR validation:

Framework unit tests run.

### CODE REVIEW

**Comment by makortel** [comment]
test parameters:
  - enable_tests = threading

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@cmsbuild, please test

Now that #35358 is merged.

**Comment by makortel** [comment]
Fine, maybe I'll just wait for the new IB.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@smuzaffar Could you remind me, is there a way to run all unit tests? Perhaps `full_cmssw = true`?

**Comment by smuzaffar** [comment]
yes @makortel , currently one way to do this is to use `full_cmssw = true` test parameter

**Comment by makortel** [comment]
test parameters:
  - enable_tests = threading
  - full_cmssw = true

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@cmsbuild, please test

One ulp old IB.

**Comment by makortel** [comment]
@Dr15Jones Could you check that the exception message looks clear?

**Comment by makortel** [comment]
@cmsbuild, please test

(refresh)

File: FWCore/Framework/interface/EventProcessor.h [hunk 1]
```diff
@@ -305,7 +305,7 @@ namespace edm {
     std::shared_ptr<EDLooperBase const> looper() const { return get_underlying_safe(looper_); }
     std::shared_ptr<EDLooperBase>& looper() { return get_underlying_safe(looper_); }
 
-    void warnAboutModulesRequiringLuminosityBLockSynchronization() const;
+    void throwAboutModulesRequiringLuminosityBLockSynchronization() const;
```
Comment (Dr15Jones) [inline FWCore/Framework/interface/EventProcessor.h:308]: Maybe fix the typo `BLock` to `Block`

**Comment by makortel** [comment]
@cmsbuild, please abort

File: FWCore/Framework/interface/EventProcessor.h (hunk 1 above)
Comment (makortel) [reply FWCore/Framework/interface/EventProcessor.h:308]:
```suggestion
    void throwAboutModulesRequiringLuminosityBlockSynchronization() const;
```

**Comment by makortel** [comment]
Nope

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
+1

@perrotta @qliphy It would be good to exercise this PR in the IBs for at least one round (mostly to really check that everything still works) before cutting the pre4. It is also ok for me to merge this PR after pre4.

**Comment by perrotta** [comment]
+1
- Try to exercize it in the 1100 IB, and see if any exception is thrown

**Pull request merged**